docs/rfcs/0041-progressive-rollout.md

```diff
@@ -0,0 +1,285 @@
+  enum Type {
+    MANUAL_SCHEDULE = 0;
+    AUTOMATIC_SCHEDULE = 1;
+  }
+  string id = 1;
+  string feature_id = 2;
+  google.protobuf.Any clause = 3;
+  Status status = 4;
+  Type type = 5;
+  int64 created_at = 6;
+  int64 updated_at = 7;
+}
+```
+
+* proto/autoops/clause.proto
+
+`ProgressiveRolloutManualScheduleClause` is set when Manual Setting is used by users.
+`ProgressiveRolloutAutomaticScheduleClause` is set when Template Setting is used by users.
+
+**NOTE**
+When `ProgressiveRolloutAutomaticScheduleClause` will be converted into `ProgressiveRollout`,
+We need to calculate `time` based on `started_at` and `interval`.
+
+```proto
+message ProgressiveRolloutScheduleClause {
+  message Schedule {
+    int64 time = 1;
+    int32 weight = 2;
+    bool triggered = 3;
+  }
+  string variation_id = 1;
+  repeated Schedule schedules = 2;
+}
+
+message ProgressiveRolloutManualScheduleClause {
+  string variation_id = 1;
+  repeated ProgressiveRolloutScheduleClause.Schedule schedules = 2;
+}
+
+message ProgressiveRolloutAutomaticScheduleClause {
+  enum Interval {
+    UNKNOWN = 0;
+    HOURLY = 1;
+    DAILY = 2;
+    WEEKLY = 3;
+  }
+  string variation_id = 1;
+  int64 started_at = 2;
+  Interval interval = 3;
+  bool triggered = 4;
+}
+```
+
+* proto/autoops/service.proto
+
+Progressive Rollout feature has unique API for creating, updating and deleting.
+
+```proto
+message CreateProgressiveRolloutRequest {
+  string environment_namespace = 1;
+  CreateProgressiveRolloutCommand command = 2;
+}
+
+message CreateProgressiveRolloutResponse {}
+
+message GetProgressiveRolloutRequest {
+  string id = 1;
+  string environment_namespace = 2;
+}
+
+message GetProgressiveRolloutResponse {
+  ProgressiveRollout progressive_rollout = 1;
+}
+
+message UpdateProgressiveRolloutRequest {
+  string id = 1;
+  string environment_namespace = 2;
+  ChangeProgressiveRolloutManualScheduleClauseCommand change_progressive_rollout_manual_schedule_clause_command = 3;
+  ChangeProgressiveRolloutAutomaticScheduleClauseCommand change_progressive_rollout_automatic_schedule_clause_command = 4;
+}
+
+message UpdateProgressiveRolloutResponse {}
+
+message DeleteProgressiveRolloutRequest {
+  string id = 1;
+  string environment_namespace = 2;
+  DeleteProgressiveRolloutCommand command = 3;
+}
+
+message DeleteProgressiveRolloutResponse {}
+
+message ListProgressiveRolloutRequest {
+  enum OrderBy {
+    DEFAULT = 0;
+    CREATED_AT = 1;
+    UPDATED_AT = 2;
+  }
+  enum OrderDirection {
+    ASC = 0;
+    DESC = 1;
+  }
+  string environment_namespace = 1;
+  int64 page_size = 2;
+  string cursor = 3;
+  repeated string feature_ids = 4;
+  OrderBy order_by = 5;
+  OrderDirection order_direction = 6;
+  google.protobuf.BoolValue is_waiting = 7;
+  google.protobuf.BoolValue is_doing = 8;
+  google.protobuf.BoolValue is_done = 9;
+  string search_keyword = 10;
+  google.protobuf.BoolValue is_manual_scheduling = 11;
+  google.protobuf.BoolValue is_automatic_scheduling = 12;
```

Comment:
IMO:

It might be better to use `status` and `type` instead of bool values like `is_waiting` and `is_manual_scheduling`.
Because there may be more `status` and `type` in the future.

It's up to you to decide to keep or change it.

Reply:
Yeah, the reason for using bool is that I'd like to define these fields as optional. However, it seems to be that `optional` field is introduced in Protocol Buffers v3.12.0.

I'll use them. Thanks